Fails to build because of `to_string`

### The problem

Hello,
I’ve just updated ESPHome Builder and attempted to rebuild my devices. I haven’t tested all of them yet, but two are now failing with similar build errors.

The configurations built successfully about 2 weeks ago on one the recent ESPHome versions. I came across a few similar issues from a couple of months back, but I wasn't able to fix my problem myself based on them (not sure if they are 100% related).

My configurations are fairly complex, so I’m not sure which parts would be most relevant to share. Is this a known issue? If so, could you suggest what areas I should focus on or what changes might have caused this?
I’d really appreciate any guidance. Thank you.

### Which version of ESPHome has the issue?

2026.2.1

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

ESP32

### Component causing the issue

_No response_

### YAML Config

```yaml

INFO ESPHome 2026.2.1
INFO Reading configuration /config/esphome/garage_door_controller.yaml...
substitutions:
  syslog_ip_address: 192.168.1.10
  home_assistant_client_name: Home Assistant
  rtttl_tone__push_button_enabled: PB_On:d=4,o=5,b=120:16g6,16c7,16p,16c7,16p,16c7
  rtttl_tone__push_button_disabled: PB_Off:d=4,o=5,b=120:16c7,16g6,16p,16g6,16p,16g6
  rtttl_tone__push_button_press_rejected: PB_Rejected:d=4,o=5,b=120:16c7,16p,16g6
  press_interval: 250ms
  event_type__garage_door_momentary_switch_button_single_pressed: ButtonSinglePressed
  event_type__garage_door_momentary_switch_button_double_pressed: ButtonDoublePressed
  event_type__garage_door_momentary_switch_button_double_pressed_and_held_down: ButtonDoublePressedAndHeldDown
  event_type__garage_door_momentary_switch_button_triple_pressed: ButtonTriplePressed
  name: garage-door-controller
  friendly_name: Garage door controller
  log_level: INFO
  component_log_level: ERROR
  garage_door:
    friendly_name: Garage door
    log_tag: door
    event_prefix: garage_door_
    opening_timeout: 30s
    closing_timeout: 30s
    momentary_switch:
      input: GPIO27
      debounce_period: 50ms
    control_unit:
      endstops:
        relay_debounce_period: 50ms
        open_indicator_relay_input: GPIO19
        closed_indicator_relay_input: GPIO21
      inputs:
        pulse_duration: 200ms
        step_by_step_relay_output: GPIO16
        partial_step_by_step_relay_output: GPIO17
  rtttl_buzzer:
    output: GPIO13
  temperature_and_humidity_sensor:
    pin: GPIO22
    min_temperature_change_to_report: '0.5'
    max_temperature_report_interval: 1h
    min_humidity_change_to_report: '1'
    max_humidity_report_interval: 1h
  board_button_pin: GPIO0
  board_smd_led_pin: GPIO23
esphome:
  on_boot:
    - priority: -100.0
      then:
        - lambda: !lambda |-
            id(global__device_initialized) = true;
  on_shutdown:
    - priority: 801.0
      then:
        - logger.log:
            format: Shutting down...
            tag: esphome
            level: INFO
            args: []
  name: garage-door-controller
  friendly_name: Garage door controller
  min_version: 2026.2.1
  build_path: build/garage-door-controller
  platformio_options: {}
  environment_variables: {}
  includes: []
  includes_c: []
  libraries: []
  name_add_mac_suffix: false
  debug_scheduler: false
  areas: []
  devices: []
esp32:
  variant: ESP32
  framework:
    type: esp-idf
    version: 5.5.2
    sdkconfig_options: {}
    log_level: ERROR
    advanced:
      compiler_optimization: SIZE
      enable_idf_experimental_features: false
      enable_lwip_assert: true
      ignore_efuse_custom_mac: false
      ignore_efuse_mac_crc: false
      enable_lwip_mdns_queries: true
      enable_lwip_bridge_interface: false
      enable_lwip_tcpip_core_locking: true
      enable_lwip_check_thread_safety: true
      disable_libc_locks_in_iram: true
      disable_vfs_support_termios: true
      disable_vfs_support_select: true
      disable_vfs_support_dir: true
      freertos_in_iram: false
      ringbuf_in_iram: false
      heap_in_iram: false
      execute_from_psram: false
      loop_task_stack_size: 8192
      enable_ota_rollback: true
      use_full_certificate_bundle: false
      include_builtin_idf_components: []
      disable_debug_stubs: true
      disable_ocd_aware: true
      disable_usb_serial_jtag_secondary: true
      disable_dev_null_vfs: true
      disable_mbedtls_peer_cert: true
      disable_mbedtls_pkcs7: true
      disable_regi2c_in_iram: true
      disable_fatfs: true
    components: []
    platform_version: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.37/platform-espressif32.zip
    source: pioarduino/framework-espidf@https://github.com/pioarduino/esp-idf/releases/download/v5.5.2/esp-idf-v5.5.2.tar.xz
  flash_size: 4MB
  board: esp32dev
  cpu_frequency: 160MHZ
wifi:
  ap:
    ssid: garage-door-controller
    password: xyz
    ap_timeout: 90s
  domain: .local
  reboot_timeout: 15min
  power_save_mode: LIGHT
  fast_connect: false
  enable_btm: false
  enable_rrm: false
  passive_scan: false
  enable_on_boot: true
  post_connect_roaming: true
  min_auth_mode: WPA2
  networks:
    - ssid: xyz
      password: xyz
      priority: 0
  use_address: garage-door-controller.local
captive_portal:
  compression: gzip
api:
  encryption:
    key: xyz
  on_client_connected:
    then:
      - logger.log:
          format: Client '%s' with an IP address %s connected to the API
          args:
            - !lambda |-
              client_info.c_str()
            - !lambda |-
              client_address.c_str()
          tag: api
          level: INFO
      - lambda: !lambda |-
          if (client_info.find("Home Assistant") == std::string::npos)
          {
            return;
          }

          id(global__homeassistant_connection_count)++;
          id(global__homeassistant_connected) = true;
          ESP_LOGI("api", "Home Assistant connected (active connections: %d)", id(global__homeassistant_connection_count));
  on_client_disconnected:
    then:
      - logger.log:
          format: Client '%s' with an IP address %s disconnected from the API
          args:
            - !lambda |-
              client_info.c_str()
            - !lambda |-
              client_address.c_str()
          tag: api
          level: INFO
      - lambda: !lambda |-
          if (client_info.find("Home Assistant") == std::string::npos)
          {
            return;
          }

          id(global__homeassistant_connection_count)--;
          id(global__homeassistant_connected) = (id(global__homeassistant_connection_count) > 0);
          ESP_LOGI("api", "Home Assistant disconnected (active connections: %d)", id(global__homeassistant_connection_count));
  actions:
    - action: rtttl_play
      variables:
        song: string
      then:
        - rtttl.play:
            id: rtttl__buzzer
            rtttl: !lambda |-
              return song;
      supports_response: none
    - action: rtttl_stop
      then:
        - rtttl.stop:
            id: rtttl__buzzer
      variables: {}
      supports_response: none
  port: 6053
  reboot_timeout: 15min
  batch_delay: 100ms
  custom_services: false
  homeassistant_services: false
  homeassistant_states: false
  listen_backlog: 4
  max_connections: 8
  max_send_queue: 8
ota:
  - platform: web_server
  - platform: esphome
    id: ota__esphome
    password: xyz
    version: 2
    port: 3232
logger:
  level: INFO
  logs:
    component: ERROR
  baud_rate: 115200
  tx_buffer_size: 512
  deassert_rts_dtr: false
  task_log_buffer_size: 768
  hardware_uart: UART0
  runtime_tag_levels: false
time:
  - platform: sntp
    id: time__sntp
    timezone: CET-1CEST,M3.5.0,M10.5.0/3
    on_time_sync:
      - then:
          - if:
              condition:
                and:
                  - time.has_time:
                      id: time__sntp
                  - lambda: !lambda |-
                      return !id(sensor__start_time_initialized);
              then:
                - component.update:
                    id: sensor__uptime_seconds
                - sensor.template.publish:
                    id: sensor__start_time
                    state: !lambda |-
                      int timestamp = id(time__sntp).utcnow().timestamp;
                      int uptime = round(id(sensor__uptime_seconds).state);
                      return timestamp - uptime;
                - globals.set:
                    id: sensor__start_time_initialized
                    value: 'true'
    update_interval: 15min
    servers:
      - 0.pool.ntp.org
      - 1.pool.ntp.org
      - 2.pool.ntp.org
udp:
  - id: udp__syslog
    addresses:
      - 192.168.1.10
    port: 18511
    listen_address: 255.255.255.255
syslog:
  udp_id: udp__syslog
  time_id: time__sntp
  port: 514
  level: INFO
  strip: true
  facility: 16
button:
  - platform: restart
    name: Garage door controller restart
    disabled_by_default: false
    icon: mdi:restart
    entity_category: config
    device_class: restart
  - platform: output
    id: button__garage_door_partial_step_by_step
    name: Garage door partial step-by-step
    icon: mdi:play-pause
    output: output__garage_door_partial_step_by_step_relay
    duration: 200ms
    on_press:
      - then:
          - logger.log:
              format: Garage door partial opening step-by-step control triggered
              tag: door
              level: DEBUG
              args: []
          - script.stop:
              id: script__garage_door_indicate_opening
          - script.stop:
              id: script__garage_door_indicate_closing
    disabled_by_default: false
  - platform: output
    id: button__garage_door_step_by_step
    name: Garage door step-by-step
    icon: mdi:play-pause
    output: output__garage_door_step_by_step_relay
    duration: 200ms
    on_press:
      - then:
          - logger.log:
              format: Garage door step-by-step control triggered
              tag: door
              level: DEBUG
              args: []
          - script.stop:
              id: script__garage_door_indicate_opening
          - script.stop:
              id: script__garage_door_indicate_closing
    disabled_by_default: false
globals:
  - id: global__device_initialized
    type: bool
    restore_value: false
    initial_value: 'false'
  - id: global__homeassistant_connection_count
    type: int
    restore_value: false
    initial_value: '0'
  - id: global__homeassistant_connected
    type: bool
    restore_value: false
    initial_value: 'false'
  - id: sensor__start_time_initialized
    type: bool
    restore_value: false
    initial_value: 'false'
sensor:
  - platform: template
    id: sensor__start_time
    name: Garage door controller start time
    device_class: timestamp
    entity_category: diagnostic
    update_interval: 4294967295ms
    disabled_by_default: false
    force_update: false
    accuracy_decimals: 1
  - platform: uptime
    id: sensor__uptime_seconds
    update_interval: 4294967295ms
    disabled_by_default: false
    force_update: false
    unit_of_measurement: s
    icon: mdi:timer-outline
    accuracy_decimals: 0
    device_class: duration
    state_class: total_increasing
    entity_category: diagnostic
    name: sensor__uptime_seconds
    internal: true
    type: seconds
  - platform: wifi_signal
    name: Garage door controller Wi-Fi RSSI
    entity_category: diagnostic
    update_interval: 60s
    on_value:
      - then:
          - lambda: !lambda "auto &signal_strength = id(sensor__wifi_signal_strength);\n\
              \nconstexpr float minRange = -100.0f;\nconstexpr float maxRange = -40.0f;\n\
              \nfloat raw_percent = (x - minRange) / (maxRange - minRange) * 100.0f;\n\
              int percent = (int) round(min(100.0f, max(0.0f, raw_percent)));\n\n\
              if (percent > 80)\n{\n  signal_strength->set_icon(\"mdi:wifi-strength-4\"\
              );\n}\nelse if (percent > 60)\n{\n  signal_strength->set_icon(\"mdi:wifi-strength-3\"\
              );\n} \nelse if (percent > 40)\n{\n  signal_strength->set_icon(\"mdi:wifi-strength-2\"\
              );\n} \nelse if (percent > 20)\n{\n  signal_strength->set_icon(\"mdi:wifi-strength-1\"\
              );\n}\nelse\n{\n  signal_strength->set_icon(\"mdi:wifi-strength-alert-outline\"\
              );\n}\n\nsignal_strength->publish_state(percent);"
    disabled_by_default: false
    force_update: false
    unit_of_measurement: dBm
    accuracy_decimals: 0
    device_class: signal_strength
    state_class: measurement
  - platform: template
    id: sensor__wifi_signal_strength
    name: Garage door controller Wi-Fi signal strength
    entity_category: diagnostic
    unit_of_measurement: '%'
    accuracy_decimals: 0
    icon: mdi:wifi-refresh
    update_interval: 4294967295ms
    disabled_by_default: false
    force_update: false
  - platform: dht
    pin:
      number: 22
      mode:
        input: true
        pullup: true
        output: false
        open_drain: false
        pulldown: false
      inverted: false
      ignore_pin_validation_error: false
      ignore_strapping_warning: false
      drive_strength: 20.0
    model: DHT22
    temperature:
      name: Garage air temperature
      accuracy_decimals: 1
      force_update: true
      filters:
        - or:
            - throttle: 1h
            - delta: 0.5
      disabled_by_default: false
      unit_of_measurement: °C
      device_class: temperature
      state_class: measurement
    humidity:
      name: Garage air humidity
      accuracy_decimals: 1
      force_update: true
      filters:
        - or:
            - throttle: 1h
            - delta: 1.0
      disabled_by_default: false
      unit_of_measurement: '%'
      device_class: humidity
      state_class: measurement
    update_interval: 60s
text_sensor:
  - platform: version
    name: Garage door controller ESPHome Version
    disabled_by_default: false
    icon: mdi:new-box
    entity_category: diagnostic
    hide_timestamp: false
binary_sensor:
  - platform: gpio
    id: binary_sensor__garage_door_is_closed
    name: Garage door is closed
    icon: mdi:garage
    pin:
      number: 21
      inverted: true
      mode:
        input: true
        pullup: true
        output: false
        open_drain: false
        pulldown: false
      ignore_pin_validation_error: false
      ignore_strapping_warning: false
      drive_strength: 20.0
    filters:
      - delayed_on_off: 50ms
    on_press:
      - then:
          - logger.log:
              format: Garage door fully closed
              tag: door
              level: DEBUG
              args: []
          - script.stop:
              id: script__garage_door_indicate_opening
          - script.stop:
              id: script__garage_door_indicate_closing
    on_release:
      - then:
          - logger.log:
              format: Garage door no longer fully closed
              tag: door
              level: DEBUG
              args: []
          - script.execute:
              id: script__garage_door_indicate_opening
    disabled_by_default: false
    use_interrupt: true
    interrupt_type: ANY
  - platform: template
    id: binary_sensor__garage_door_is_closing
    name: Garage door is closing
    device_class: moving
    icon: mdi:arrow-down
    lambda: !lambda |-
      return id(script__garage_door_indicate_closing).is_running();
    on_press:
      - then:
          - script.execute:
              id: script__garage_door_publish_cover_operation
              operation: COVER_OPERATION_CLOSING
    on_release:
      - then:
          - script.execute:
              id: script__garage_door_publish_cover_operation
              operation: COVER_OPERATION_IDLE
    disabled_by_default: false
  - platform: gpio
    name: Garage door momentary switch pressed
    icon: mdi:gesture-tap
    pin:
      number: 27
      inverted: true
      mode:
        input: true
        pullup: true
        output: false
        open_drain: false
        pulldown: false
      ignore_pin_validation_error: false
      ignore_strapping_warning: false
      drive_strength: 20.0
    filters:
      - delayed_on_off: 50ms
    on_press:
      - then:
          - logger.log:
              format: Garage door momentary switch pressed
              tag: door
              level: DEBUG
              args: []
    on_release:
      - then:
          - logger.log:
              format: Garage door momentary switch released
              tag: door
              level: DEBUG
              args: []
    on_multi_click:
      - timing:
          - state: true
            min_length: 0ms
            max_length: 1s
          - state: false
            min_length: 250ms
        then:
          - script.execute:
              id: momentary_switch_on_single_press
        invalid_cooldown: 1s
      - timing:
          - state: true
            min_length: 0ms
            max_length: 250ms
          - state: false
            min_length: 0ms
            max_length: 250ms
          - state: true
            min_length: 0ms
            max_length: 250ms
          - state: false
            min_length: 250ms
        then:
          - script.execute:
              id: momentary_switch_on_double_press
        invalid_cooldown: 1s
      - timing:
          - state: true
            min_length: 0ms
            max_length: 250ms
          - state: false
            min_length: 0ms
            max_length: 250ms
          - state: true
            min_length: 0ms
            max_length: 250ms
          - state: false
            min_length: 0ms
            max_length: 250ms
          - state: true
            min_length: 0ms
            max_length: 250ms
          - state: false
            min_length: 250ms
        then:
          - script.execute:
              id: momentary_switch_on_triple_press
        invalid_cooldown: 1s
      - timing:
          - state: true
            min_length: 0ms
            max_length: 250ms
          - state: false
            min_length: 0ms
            max_length: 250ms
          - state: true
            min_length: 0ms
            max_length: 250ms
          - state: false
            min_length: 0ms
            max_length: 250ms
          - state: true
            min_length: 2s
        then:
          - script.execute:
              id: momentary_switch_on_double_press_and_hold_down
        invalid_cooldown: 1s
    disabled_by_default: false
    use_interrupt: true
    interrupt_type: ANY
  - platform: gpio
    id: binary_sensor__garage_door_is_open
    name: Garage door is open
    device_class: garage_door
    pin:
      number: 19
      inverted: true
      mode:
        input: true
        pullup: true
        output: false
        open_drain: false
        pulldown: false
      ignore_pin_validation_error: false
      ignore_strapping_warning: false
      drive_strength: 20.0
    filters:
      - delayed_on_off: 50ms
    on_press:
      - then:
          - logger.log:
              format: Garage door fully open
              tag: door
              level: DEBUG
              args: []
          - script.stop:
              id: script__garage_door_indicate_opening
          - script.stop:
              id: script__garage_door_indicate_closing
    on_release:
      - then:
          - logger.log:
              format: Garage door no longer fully open
              tag: door
              level: DEBUG
              args: []
          - script.execute:
              id: script__garage_door_indicate_closing
    disabled_by_default: false
    use_interrupt: true
    interrupt_type: ANY
  - platform: template
    id: binary_sensor__garage_door_is_opening
    name: Garage door is opening
    device_class: moving
    icon: mdi:arrow-up
    lambda: !lambda |-
      return id(script__garage_door_indicate_opening).is_running();
    on_press:
      - then:
          - script.execute:
              id: script__garage_door_publish_cover_operation
              operation: COVER_OPERATION_OPENING
    on_release:
      - then:
          - script.execute:
              id: script__garage_door_publish_cover_operation
              operation: COVER_OPERATION_IDLE
    disabled_by_default: false
cover:
  - platform: template
    id: cover__garage_door_step_by_step
    name: Garage door step-by-step
    device_class: garage
    optimistic: false
    restore_mode: NO_RESTORE
    has_position: true
    lambda: !lambda |-
      constexpr float COVER_HALF_OPEN = 0.5f;

      const bool is_open = id(binary_sensor__garage_door_is_open).state;
      const bool is_closed = id(binary_sensor__garage_door_is_closed).state;

      // INVALID: both end-stops active
      if (is_open && is_closed)
      {
        return COVER_HALF_OPEN;
      }

      // OPEN
      if (is_open)
      {
        return COVER_OPEN;
      }

      // CLOSED
      if (is_closed)
      {
        return COVER_CLOSED;
      }

      // PARTIAL
      return COVER_HALF_OPEN;
    open_action:
      then:
        - if:
            condition:
              lambda: !lambda |-
                return
                  !id(binary_sensor__garage_door_is_open).state &&
                  !id(binary_sensor__garage_door_is_opening).state;
            then:
              - button.press:
                  id: button__garage_door_step_by_step
    close_action:
      then:
        - if:
            condition:
              lambda: !lambda |-
                return
                  !id(binary_sensor__garage_door_is_closed).state &&
                  !id(binary_sensor__garage_door_is_closing).state;
            then:
              - button.press:
                  id: button__garage_door_step_by_step
    stop_action:
      then:
        - if:
            condition:
              lambda: !lambda |-
                return
                  id(binary_sensor__garage_door_is_opening).state ||
                  id(binary_sensor__garage_door_is_closing).state;
            then:
              - button.press:
                  id: button__garage_door_step_by_step
    toggle_action:
      then:
        - button.press:
            id: button__garage_door_step_by_step
    disabled_by_default: false
    assumed_state: false
  - platform: template
    id: cover__garage_door_partial_step_by_step
    name: Garage door partial step-by-step
    device_class: garage
    optimistic: false
    restore_mode: NO_RESTORE
    has_position: true
    lambda: !lambda |-
      return id(cover__garage_door_step_by_step).position;
    open_action:
      then:
        - if:
            condition:
              lambda: !lambda |-
                return
                  !id(binary_sensor__garage_door_is_open).state &&
                  !id(binary_sensor__garage_door_is_opening).state;
            then:
              - button.press:
                  id: button__garage_door_partial_step_by_step
    close_action:
      then:
        - if:
            condition:
              lambda: !lambda |-
                return
                  !id(binary_sensor__garage_door_is_closed).state &&
                  !id(binary_sensor__garage_door_is_closing).state;
            then:
              - button.press:
                  id: button__garage_door_partial_step_by_step
    stop_action:
      then:
        - if:
            condition:
              lambda: !lambda |-
                return
                  id(binary_sensor__garage_door_is_opening).state ||
                  id(binary_sensor__garage_door_is_closing).state;
            then:
              - button.press:
                  id: button__garage_door_partial_step_by_step
    toggle_action:
      then:
        - button.press:
            id: button__garage_door_partial_step_by_step
    disabled_by_default: false
    assumed_state: false
event:
  - platform: template
    id: event__garage_door_momentary_switch
    name: Garage door momentary switch
    icon: mdi:gesture-tap
    device_class: button
    event_types:
      - ButtonSinglePressed
      - ButtonDoublePressed
      - ButtonDoublePressedAndHeldDown
      - ButtonTriplePressed
    on_event:
      - then:
          - logger.log:
              format: Event '%s' triggered
              tag: door
              level: DEBUG
              args:
                - !lambda |-
                  event_type.c_str()
          - homeassistant.event:
              event: esphome.garage_door_momentary_switch_event
              data:
                event_name: !lambda |-
                  return event_type;
              data_template: {}
              variables: {}
    disabled_by_default: false
output:
  - platform: gpio
    id: output__garage_door_partial_step_by_step_relay
    pin:
      number: 17
      inverted: false
      mode:
        output: true
        input: false
        open_drain: false
        pullup: false
        pulldown: false
      ignore_pin_validation_error: false
      ignore_strapping_warning: false
      drive_strength: 20.0
  - platform: ledc
    id: output__rtttl_buzzer
    pin:
      number: 13
      mode:
        output: true
        input: false
        open_drain: false
        pullup: false
        pulldown: false
      inverted: false
      ignore_pin_validation_error: false
      ignore_strapping_warning: false
      drive_strength: 20.0
    zero_means_zero: false
    frequency: 1000.0
  - platform: gpio
    id: output__garage_door_step_by_step_relay
    pin:
      number: 16
      inverted: false
      mode:
        output: true
        input: false
        open_drain: false
        pullup: false
        pulldown: false
      ignore_pin_validation_error: false
      ignore_strapping_warning: false
      drive_strength: 20.0
rtttl:
  - id: rtttl__buzzer
    output: output__rtttl_buzzer
    gain: 0.6
script:
  - id: script__garage_door_indicate_closing
    mode: single
    then:
      - script.stop:
          id: script__garage_door_indicate_opening
      - delay: 30s
    parameters: {}
  - id: script__garage_door_indicate_opening
    mode: single
    then:
      - script.stop:
          id: script__garage_door_indicate_closing
      - delay: 30s
    parameters: {}
  - id: script__garage_door_publish_cover_operation
    parameters:
      operation: CoverOperation
    then:
      - cover.template.publish:
          id: cover__garage_door_step_by_step
          current_operation: !lambda |-
            return operation;
      - cover.template.publish:
          id: cover__garage_door_partial_step_by_step
          current_operation: !lambda |-
            return operation;
    mode: single
  - id: momentary_switch_on_double_press_and_hold_down
    then:
      - logger.log:
          format: Garage door momentary switch double-pressed and held down (triggering
            step-by-step control regardless of momentary switch protection)
          tag: door
          level: DEBUG
          args: []
      - button.press:
          id: button__garage_door_step_by_step
      - event.trigger:
          id: event__garage_door_momentary_switch
          event_type: ButtonDoublePressedAndHeldDown
    mode: single
    parameters: {}
  - id: momentary_switch_on_double_press
    then:
      - if:
          condition:
            and:
              - switch.is_on:
                  id: switch__garage_door_momentary_switch_enabled
          then:
            - logger.log:
                format: Garage door momentary switch double-pressed (triggering partial
                  step-by-step control)
                tag: door
                level: DEBUG
                args: []
            - button.press:
                id: button__garage_door_partial_step_by_step
          else:
            - logger.log:
                format: Garage door momentary switch double press ignored because
                  the switch is disabled (partial step-by-step control will not be
                  triggered)
                tag: door
                level: DEBUG
                args: []
            - rtttl.play:
                id: rtttl__buzzer
                rtttl: PB_Rejected:d=4,o=5,b=120:16c7,16p,16g6
      - event.trigger:
          id: event__garage_door_momentary_switch
          event_type: ButtonDoublePressed
    mode: single
    parameters: {}
  - id: momentary_switch_on_single_press
    then:
      - if:
          condition:
            and:
              - switch.is_on:
                  id: switch__garage_door_momentary_switch_enabled
          then:
            - logger.log:
                format: Garage door momentary switch single-pressed (triggering step-by-step
                  control)
                tag: door
                level: DEBUG
                args: []
            - button.press:
                id: button__garage_door_step_by_step
          else:
            - logger.log:
                format: Garage door momentary switch single press ignored because
                  the switch is disabled (step-by-step control will not be triggered)
                tag: door
                level: DEBUG
                args: []
            - rtttl.play:
                id: rtttl__buzzer
                rtttl: PB_Rejected:d=4,o=5,b=120:16c7,16p,16g6
      - event.trigger:
          id: event__garage_door_momentary_switch
          event_type: ButtonSinglePressed
    mode: single
    parameters: {}
  - id: momentary_switch_on_triple_press
    then:
      - if:
          condition:
            and:
              - switch.is_on:
                  id: switch__garage_door_momentary_switch_enabled
          then:
            - logger.log:
                format: Garage door momentary switch triple-pressed (disabling switch)
                tag: door
                level: DEBUG
                args: []
            - switch.turn_off:
                id: switch__garage_door_momentary_switch_enabled
            - rtttl.play:
                id: rtttl__buzzer
                rtttl: PB_Off:d=4,o=5,b=120:16c7,16g6,16p,16g6,16p,16g6
          else:
            - logger.log:
                format: Garage door momentary switch triple-pressed (enabling switch)
                tag: door
                level: DEBUG
                args: []
            - switch.turn_on:
                id: switch__garage_door_momentary_switch_enabled
            - rtttl.play:
                id: rtttl__buzzer
                rtttl: PB_On:d=4,o=5,b=120:16g6,16c7,16p,16c7,16p,16c7
      - event.trigger:
          id: event__garage_door_momentary_switch
          event_type: ButtonTriplePressed
    mode: single
    parameters: {}
switch:
  - platform: template
    id: switch__garage_door_momentary_switch_enabled
    name: Garage door momentary switch enabled
    icon: mdi:gesture-tap
    restore_mode: ALWAYS_ON
    optimistic: true
    turn_on_action:
      then:
        - logger.log:
            format: Momentary switch enabled
            tag: door
            level: DEBUG
            args: []
    turn_off_action:
      then:
        - logger.log:
            format: Momentary switch disabled
            tag: door
            level: DEBUG
            args: []
    disabled_by_default: false
    assumed_state: false
preferences:
  flash_write_interval: 6h

```

### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2026.2.1
INFO Reading configuration /config/esphome/garage_door_controller.yaml...
INFO Generating C++ source...
INFO Setting CONFIG_LWIP_MAX_SOCKETS to 13 (registered: api=4, captive_portal=4, mdns=2, ota=1, udp=2)
INFO Compiling app... Build path: /data/build/garage-door-controller
Processing garage-door-controller (board: esp32dev; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.37/platform-espressif32.zip)
--------------------------------------------------------------------------------
Library Manager: Installing esphome/noise-c @ 0.1.10
INFO Installing esphome/noise-c @ 0.1.10
Unpacking  [####################################]  100%
Library Manager: noise-c@0.1.10 has been installed!
INFO noise-c@0.1.10 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/libsodium @ 1.10020.7
INFO Installing esphome/libsodium @ 1.10020.7
Unpacking  [####################################]  100%
Library Manager: libsodium@1.10020.7 has been installed!
INFO libsodium@1.10020.7 has been installed!
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - contrib-piohome @ 3.4.4 
 - framework-espidf @ 3.50502.0 (5.5.2) 
 - tool-cmake @ 4.0.3 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.1.2 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-xtensa-esp-elf @ 14.2.0+20251107
Reading CMake configuration...
Generating assembly for certificate bundle...
Dependency Graph
|-- noise-c @ 0.1.10
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/api_frame_helper_noise.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/api/user_services.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/binary_sensor/automation.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/binary_sensor/binary_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/binary_sensor/filter.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/button/button.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/captive_portal/dns_server_esp32_idf.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/cover/cover.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/dht/dht.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/esp32/core.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/esp32/gpio.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/esp32/helpers.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/esp32/preferences.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/esp32/throw_stubs.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/esphome/ota/ota_esphome.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/event/event.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/gpio/binary_sensor/gpio_binary_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/gpio/output/gpio_binary_output.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/ledc/ledc_output.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/logger/logger_esp32.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/logger/task_log_buffer_esp32.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/logger/task_log_buffer_zephyr.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/mdns/mdns_esp32.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/ota/ota_backend.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/ota/ota_backend_host.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/output/automation.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/output/button/output_button.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/output/float_output.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/restart/button/restart_button.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/rtttl/rtttl.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/safe_mode/safe_mode.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/script/script.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/sensor/automation.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/sensor/filter.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/sensor/sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/sha256/sha256.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/sntp/sntp_component.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/socket/bsd_sockets_impl.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/socket/socket.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/switch/automation.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/switch/switch.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/syslog/esphome_syslog.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/template/binary_sensor/template_binary_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/template/cover/template_cover.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/template/sensor/template_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/template/switch/template_switch.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/text_sensor/filter.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/text_sensor/text_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/time/automation.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/time/real_time_clock.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/udp/udp_component.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/uptime/sensor/uptime_seconds_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/uptime/sensor/uptime_timestamp_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/version/version_text_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/web_server/ota/ota_web_server.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/web_server_base/web_server_base.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/web_server_idf/multipart.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/web_server_idf/utils.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/web_server_idf/web_server_idf.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/wifi/wifi_component_esp_idf.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/components/wifi_signal/wifi_signal_sensor.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/application.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/color.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/component.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/component_iterator.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/controller_registry.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/entity_base.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/gpio.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/log.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/ring_buffer.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/time.cpp.o
Compiling .pioenvs/garage-door-controller/src/esphome/core/util.cpp.o
Compiling .pioenvs/garage-door-controller/src/main.cpp.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_core/ed25519/ref10/ed25519_ref10.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_core/salsa/ref/core_salsa_ref.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_hash/crypto_hash.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_hash/sha256/cp/hash_sha256_cp.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_hash/sha256/hash_sha256.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_onetimeauth/crypto_onetimeauth.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_onetimeauth/poly1305/donna/poly1305_donna.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_onetimeauth/poly1305/onetimeauth_poly1305.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_scalarmult/crypto_scalarmult.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_scalarmult/curve25519/ref10/x25519_ref10.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_scalarmult/curve25519/scalarmult_curve25519.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/chacha20/ref/chacha20_ref.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/chacha20/stream_chacha20.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/crypto_stream.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/salsa20/ref/salsa20_ref.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/salsa20/stream_salsa20.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/salsa20/xmm6/salsa20_xmm6-asm.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/salsa20/xmm6/salsa20_xmm6.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/salsa20/xmm6int/salsa20_xmm6int-avx2.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_stream/salsa20/xmm6int/salsa20_xmm6int-sse2.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/crypto_verify/verify.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/randombytes/internal/randombytes_internal_random.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/randombytes/randombytes.o
In file included from src/esphome.h:13,
                 from src/main.cpp:3:
src/esphome/components/api/homeassistant_service.h: In instantiation of 'static std::string esphome::api::TemplatableStringValue<X>::value_to_string(T&&) [with T = esphome::StringRef; X = {esphome::StringRef}; std::string = std::__cxx11::basic_string<char>]':
src/esphome/components/api/homeassistant_service.h:48:96:   required from 'esphome::api::TemplatableStringValue<X>::TemplatableStringValue(F) [with F = setup()::<lambda(esphome::StringRef)>; typename std::enable_if<std::is_invocable<F, X ...>::value, int>::type <anonymous> = 0; X = {esphome::StringRef}]'
   48 |       : TemplatableValue<std::string, X...>([f](X... x) -> std::string { return value_to_string(f(x...)); }) {}
      |                                                                                 ~~~~~~~~~~~~~~~^~~~~~~~~
src/esphome/components/api/homeassistant_service.h:214:14:   required from 'void esphome::api::HomeAssistantServiceCallAction<Ts>::add_kv_(esphome::FixedVector<esphome::api::TemplatableKeyValuePair<Ts ...> >&, const char*, V&&) [with V = setup()::<lambda(esphome::StringRef)>; Ts = {esphome::StringRef}]'
  214 |     kv.value = std::forward<V>(value);
      |     ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/api/homeassistant_service.h:122:18:   required from 'void esphome::api::HomeAssistantServiceCallAction<Ts>::add_data(const char*, V&&) [with V = setup()::<lambda(esphome::StringRef)>; Ts = {esphome::StringRef}]'
  122 |     this->add_kv_(this->data_, key, std::forward<V>(value));
      |     ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/config/esphome/garage_door_controller/covers/_stop_action.yaml:69:50:   required from here
src/esphome/components/api/homeassistant_service.h:29:21: error: no matching function for call to 'to_string(esphome::StringRef)'
   29 |     return to_string(std::forward<T>(val));  // NOLINT
      |            ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
In file included from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/string:54,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/locale_classes.h:40,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/ios_base.h:41,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/ios:44,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/ostream:40,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/unique_ptr.h:43,
                 from /data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/memory:78,
                 from src/esphome/components/api/api_frame_helper.h:5,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3:
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4421:3: note: candidate: 'std::string std::__cxx11::to_string(long double)'
 4421 |   to_string(long double __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4421:25: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'long double'
 4421 |   to_string(long double __val)
      |             ~~~~~~~~~~~~^~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4411:3: note: candidate: 'std::string std::__cxx11::to_string(double)'
 4411 |   to_string(double __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4411:20: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'double'
 4411 |   to_string(double __val)
      |             ~~~~~~~^~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4401:3: note: candidate: 'std::string std::__cxx11::to_string(float)'
 4401 |   to_string(float __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4401:19: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'float'
 4401 |   to_string(float __val)
      |             ~~~~~~^~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4327:3: note: candidate: 'std::string std::__cxx11::to_string(long long unsigned int)'
 4327 |   to_string(unsigned long long __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4327:32: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'long long unsigned int'
 4327 |   to_string(unsigned long long __val)
      |             ~~~~~~~~~~~~~~~~~~~^~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4310:3: note: candidate: 'std::string std::__cxx11::to_string(long long int)'
 4310 |   to_string(long long __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4310:23: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'long long int'
 4310 |   to_string(long long __val)
      |             ~~~~~~~~~~^~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4294:3: note: candidate: 'std::string std::__cxx11::to_string(long unsigned int)'
 4294 |   to_string(unsigned long __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4294:27: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'long unsigned int'
 4294 |   to_string(unsigned long __val)
      |             ~~~~~~~~~~~~~~^~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4275:3: note: candidate: 'std::string std::__cxx11::to_string(long int)'
 4275 |   to_string(long __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4275:18: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'long int'
 4275 |   to_string(long __val)
      |             ~~~~~^~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4259:3: note: candidate: 'std::string std::__cxx11::to_string(unsigned int)'
 4259 |   to_string(unsigned __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4259:22: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'unsigned int'
 4259 |   to_string(unsigned __val)
      |             ~~~~~~~~~^~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4240:3: note: candidate: 'std::string std::__cxx11::to_string(int)'
 4240 |   to_string(int __val)
      |   ^~~~~~~~~
/data/cache/platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/14.2.0/bits/basic_string.h:4240:17: note:   no known conversion for argument 1 from 'esphome::StringRef' to 'int'
 4240 |   to_string(int __val)
      |             ~~~~^~~~~
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/randombytes/sysrandom/randombytes_sysrandom.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/sodium/codecs.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/sodium/core.o
Compiling .pioenvs/garage-door-controller/libbb3/libsodium/sodium/runtime.o
*** [.pioenvs/garage-door-controller/src/main.cpp.o] Error 1
========================= [FAILED] Took 187.01 seconds =========================
```

### Additional information

_No response_

--- a/esphome/components/api/homeassistant_service.h
+++ b/esphome/components/api/homeassistant_service.h
@@ -36,6 +36,8 @@ template<typename... X> class TemplatableStringValue : public TemplatableValue<s
   static std::string value_to_string(const char *val) { return std::string(val); }  // For lambdas returning .c_str()
   static std::string value_to_string(const std::string &val) { return val; }
   static std::string value_to_string(std::string &&val) { return std::move(val); }
+  static std::string value_to_string(const StringRef &val) { return val.str(); }
+  static std::string value_to_string(StringRef &&val) { return val.str(); }
 
  public:
   TemplatableStringValue() : TemplatableValue<std::string, X...>() {}

--- a/tests/components/homeassistant/common.yaml
+++ b/tests/components/homeassistant/common.yaml
@@ -90,6 +90,19 @@ text_sensor:
     id: ha_hello_world_text2
     attribute: some_attribute
 
+event:
+  - platform: template
+    name: Test Event
+    id: test_event
+    event_types:
+      - test_event_type
+    on_event:
+      - homeassistant.event:
+          event: esphome.test_event
+          data:
+            event_name: !lambda |-
+              return event_type;
+
 time:
   - platform: homeassistant
     on_time:

